improve: generate crds in parallel and optimize code

nicoloboschi opened the pull request:
## Description
In my project I have 10 CRDs classes. 
I've taken a CPU flamegraph and I figured out the usage of java streams api adds a relevant overhead while visiting specs/status classes via reflection. 

Also the visitors can access the crd in parallel since the base typeref is stateless.

Changes included:
- Enable concurrent crd visitors using nCPU threads.
- Moved some hot path code from using streams to old plain java for-loop

FYI I did a similar fix in sundrio: https://github.com/sundrio/sundrio/pull/352

The net result of this patch is that generation time moved from ~30 seconds to to ~11 seconds.  

## Type of change
<!---
What types of changes does your code introduce? Put an `x` in all the boxes that apply
-->
 - [x] Bug fix (non-breaking change which fixes an issue)
 - [ ] Feature (non-breaking change which adds functionality)
 - [ ] Breaking change (fix or feature that would cause existing functionality to change
 - [ ] Chore (non-breaking change which doesn't affect codebase;
   test, version modification, documentation, etc.)

## Checklist
 - [x] Code contributed by me aligns with current project license: [Apache 2.0](https://www.apache.org/licenses/LICENSE-2.0)
 - [x] I Added [CHANGELOG](https://github.com/fabric8io/kubernetes-client/blob/master/CHANGELOG.md) entry regarding this change
 - [x] I have implemented unit tests to cover my changes
 - [x] I have added/updated the [javadocs](https://www.javadoc.io/doc/io.fabric8/kubernetes-client/latest/index.html) and other [documentation](https://github.com/fabric8io/kubernetes-client/blob/master/doc/CHEATSHEET.md) accordingly
 - [x] No new bugs, code smells, etc. in [SonarCloud](https://sonarcloud.io/dashboard?id=fabric8io_kubernetes-client) report
 - [ ] I tested my code in Kubernetes
 - [ ] I tested my code in OpenShift

<!--
Integration tests (https://github.com/fabric8io/kubernetes-client/tree/master/kubernetes-itests)
Please check integration tests and provide/improve tests if applicable.

Open your PR in Draft mode and verify all of the applicable Checklist items before marking your pull request as ready for review
-->

andreaTP commented:
Hi @nicoloboschi , thanks for this PR! This sounds super interesting indeed!
But, since I have already experienced issues with concurrency + sundrio + crd-gen, I need to summon @metacosm and @iocanel to review this change 🙏

andreaTP commented:
@nicoloboschi are you still actively working on this? if that's the case, would you mind converting to Draft?

nicoloboschi commented on crd-generator/api/src/main/java/io/fabric8/crd/generator/AbstractCustomResourceHandler.java:
```diff
@@ -54,21 +60,53 @@ public void handle(CustomResourceInfo config) {
-      builder.accept(specReplicasPathDetector);
-    }
+    final ExecutorService executorService = Executors.newFixedThreadPool(
+      Math.min(5, Runtime.getRuntime().availableProcessors()));
```
They are not going to compete for any resource because they are all independent tasks (that's what I've seen but I don't know this code very deeply). 

I haven't measured the cache hit ratio but the results show the parallel execution is drastically faster.

iocanel reviewed (approved):
Overall, looks good.
Added a comment with two suggestion, which I consider nice ot have.

nicoloboschi commented on crd-generator/api/src/main/java/io/fabric8/crd/generator/AbstractCustomResourceHandler.java:
```diff
@@ -54,21 +60,51 @@ public void handle(CustomResourceInfo config) {
+      List<CompletableFuture<Void>> futures = new ArrayList<>();
+      if (config.specClassName().isPresent()) {
+        futures.add(CompletableFuture.runAsync(() -> {
+          TypeDefBuilder builder = new TypeDefBuilder(def);
```
>However, it shouldn't make a difference as in this particular case we are not interested in building something (we completely ignore that), but instead we just need to traverse the object graph.

yes, exactly, we just do a LOT of recursive calls and the only output is stored in the visitor object itself (`getPath()`)

I've already tried with the var-args solution but it doesn't help.

Sharing my dummy benchmark numbers
- using quarkus-operator-sdk mojo to generate crd
- jdk 17
- Intel mac 6 core
- 8 CRD in the project with ~200 props in total, spec and status classes in each one
- 3 rounds per solution
Results:
- current version: ~40s
- concurrency of 5: ~18s
- var args:  ~45s

nicoloboschi commented:
@iocanel any chance we can merge this? 

this is my latest comment about performances comparison


- using quarkus-operator-sdk mojo to generate crd
- jdk 17
- Intel mac 6 core
- 8 CRD in the project with ~200 props in total, spec and status classes in each one
- 3 rounds per solution

Results:
- current version: ~40s
- concurrency of 5: ~18s
- var args: ~45s

nicoloboschi commented:
@iocanel @andreaTP any chance to see this pull moving forward? 🤞

manusa commented:
Hi @nicoloboschi 
I still need to review this and analyze if it might cause trouble in certain environments. AFAIK running the generation in several parallel processes might lead to inconsistent results.
One of the options is to make this a configuration option.

nicoloboschi commented:
> Hi @nicoloboschi I still need to review this and analyze if it might cause trouble in certain environments. AFAIK running the generation in several parallel processes might lead to inconsistent results. One of the options is to make this a configuration option.

@manusa I can totally add a flag. I haven't find any configuration present at the moment. Do you have any suggestion on how to make it configurable? 

For example, I could add a system property but I'm not sure how it is supposed to be set if someone is using the `crd-generator-apt` module

andreaTP commented:
> Do you have any suggestion on how to make it configurable?

Currently, the crd generator is purely using annotations and has no additional configuration:
https://github.com/fabric8io/kubernetes-client/tree/009974fe5fe634cbd6d1210780b99ff82e251f2d/crd-generator/api/src/main/java/io/fabric8/crd/generator/annotation

nicoloboschi commented:
- I've made it configurable and disabled by default.
- I added an option in `CRDGenerator` for enabling the parallel execution. 
- For the apt processor I've added an option `io.fabric8.crd.generator.parallel`. 
- I added the apt processor option example in the test pom.xml
```
<compilerArgs>
            <arg>-Aio.fabric8.crd.generator.parallel=false</arg>
</compilerArgs>
```
- I've added CRDGenerator tests with parallel enabled
- I tested locally the `crd-generator/tests` suite with parallel enabled but I've left it disabled since it's better to test the default configuration in this case

@andreaTP let me know if I missed something

andreaTP commented:
I just skimmed through on the phone, and, making parallel execution optional makes a lot of sense.
I think that the Maven setting is a bit "hidden" and you probably want to mention it in the docs.

I might be wrong, but it looks like there are a few tiny bits that are not source compatible (which is appropriate to merge before 6.4) and @metacosm would need to have a look and follow up in the Quarkus Operator SDK.

nicoloboschi commented:
I created the related pr in for the quarkus sdk: https://github.com/quarkiverse/quarkus-operator-sdk/pull/488 (it's only a draft since we need this to get released)

andreaTP commented on crd-generator/test/pom.xml:
```diff
@@ -90,6 +90,14 @@
+        <artifactId>maven-compiler-plugin</artifactId>
+        <configuration>
+          <compilerArgs>
+            <arg>-Aio.fabric8.crd.generator.parallel=false</arg>
```
this should be redundant as it's already the default, isn't it?

nicoloboschi replied:
yes it is. I wanted to make it explicit and leave it there as example since it's not trivial to add processor options

andreaTP replied:
👍

andreaTP commented:
Thanks for the follow-ups @nicoloboschi , LGTM.

metacosm reviewed (approved):
LGTM  but I haven't checked downstream projects yet.

gastaldi commented on crd-generator/api/src/main/java/io/fabric8/crd/generator/AbstractCustomResourceHandler.java:
```diff
@@ -91,6 +98,54 @@ public void handle(CustomResourceInfo config) {
+  }
+
+  private TypeDef visitTypeDefInParallel(TypeDef def, List<Visitor<TypeDefBuilder>> visitors) {
+    if (visitors.isEmpty()) {
+      return def;
+    }
```
Perhaps move this `if` to the `visitTypeDef` method since it is useful for `visitTypeDefSequentially` too?

nicoloboschi replied:
done

gastaldi commented on crd-generator/api/src/main/java/io/fabric8/crd/generator/AbstractCustomResourceHandler.java:
```diff
@@ -91,6 +98,54 @@ public void handle(CustomResourceInfo config) {
+        }, executorService));
+      }
+      try {
+        CompletableFuture.allOf(futures.toArray(new CompletableFuture[] {})).get();
```
```suggestion
        CompletableFuture.allOf(futures.toArray(new CompletableFuture[0])).get();
```

nicoloboschi commented:
@gastaldi I've addressed your comments, please review it again

gastaldi commented on doc/CRD-generator.md:
```diff
@@ -442,3 +442,32 @@ The CRD generator will add the additional `labels`:
+## Experimental
+
+### Generate CRDs in parallel
+Starting from 6.4, it's possible to speed up the CRDs generation by using parallel computation.
```
6.4.0 is already released, since this change doesn't break the API it could be introduced in a patch 6.4.x release

gastaldi commented on crd-generator/api/src/main/java/io/fabric8/crd/generator/CRDGenerator.java:
```diff
@@ -68,6 +69,11 @@ public CRDGenerator withOutput(CRDOutput<? extends OutputStream> output) {
     return this;
   }
 
+  public CRDGenerator withParallelGenerationEnabled(boolean parallel) {
```
Just an idea: what if the ExecutorService is passed instead? That would give a better control over the execution.

nicoloboschi replied:
In general I think this makes sense. 
In this specific case, given the crd generation is a short-running task, I don't believe it's worth to manage the thread pool at higher level

gastaldi reviewed (approved):
LGTM. I'd appreciate if you could please squash the commits. Good job!

nicoloboschi commented:
Fixed the doc, rebased to the current master and squashed @gastaldi

manusa reviewed (approved):
Awesome work, thx!
